### ISSUE TITLE
Please discuss: Splitting personal names?

### ISSUE THREAD

**Issue by raphaelm** [pretix/pretix#978]
## Current situation

pretix has two dedicated name fields, ``OrderPosition.attendee_name`` and ``InvoiceAddress.name``. Both are just a single text fields that store a full name.

The reasoning behind storing names as a single string is that it is ready for full internationalization. Names are structured very differently across cultures and building cultural assumptions into the database design sounds like a bad idea. 

However, there's a strong urge by many of our users to split the field into multiple ones. Most of the times, it is requested that we split it into "first name" and "last name", sometimes also a "title" field is asked for.

## Use cases

Before we go any further, we should define what people need split names for. So far, I've heard the following:

* **Sorting.** In a business context in most Western cultures, you'll want to be able to sort attendee lists etc. by last name; while in an informal context, you might want to sort by first name. It get's more complicated, since e.g. the Spanish name "María-Jose Carreño Quiñones" would be sorted by C in some Spanish-speaking countries, and by Q in others. The same goes for prefixes and suffixes like "von" in German, although everyone with a "von" in the name is probably used to being sorted at "v" from time to time, so we could get away with that.
* **Salutations.** In emails, it might be desireable to address people by their first or last name specifically.
* **Badge design.** On Badges, you might want to omit parts of the name, or print parts of the name in bold, etc.
* **Cross-referencing** with external data stores.

If you can think of any other use cases for which you need names split, please let us know.

## Research

The W3C has a recommendation on how to handle names: https://www.w3.org/International/questions/qa-personal-names.en

Generally, it recommends to ask for a "full name" and a "calling name", if you just want to solve the **salutation** use case. However, this doesn't help at all for sorting.

In that case, they prefer to go with "family name" and "given name" instead of "first name" and "last name", as the order of "family name" and "given name" varies between cultures, whereas most (not all!) cultures at least have the concept of family names and given names.

Some things, like Japanese names which cannot be sorted without knowing their latin description, are even harder, but we probably can't solve it for all cultures.

## Proposed design A: Hardcode a simple semi-Western scheme

Split the field into "family name", "given name" and "salutation", which can be used for a title or something else, and provide sensible defaults for the salutation. Allow the user to sort by any of these fields and expose them individually to badge/ticket designs.

Upsides:
* Easily implemented.

Downsides:
* This is bad for internationalization.
* We need to figure out how to put names together for display. Some people will want "Family name, Given name" and some will want "Given Name, Family name" and some people will want it completely differen.

## Proposed design B: Fixed number of configurable fields

Split the field into 4(?) fields, "name part 1", "name part 2", "name part 3" and "name part 4", and introduce event-level config options that allow the user to decide how many fields they want and how they should be labeled. Allow the user to sort by any of these fields and expose them individually to badge/ticket designs.

The above scheme could become a default, instead of a forced requirement.

Upsides:
* Allowes for pretty flexible schemes in a large number of cultures, and also leaves the freedom to e.g. only ask for given names or omit salutations etc.

Downsides:
* This is bad database design, since there's no semantic meaning to the columns and there's an arbitrary limitation of 4(?) fields. Creating a new database table just to store name parts seems overkill as well and will make querying much harder – or doesn't it?
* We need to figure out how to put names together for display. Some people will want "name part 3, name part 2" and some will want "name part 2 name part 3" and some people will want it completely differen.

## More designs?

We're happy for more ideas on how to do this!

## Things that will get harder

* Search will get harder: With one-field names, you can just do a substring search to perform a query, with split  names, we need to build more complex queries to find people even if the search query is a full name.
* When buying an add-on product, the attendee name of the add-on will be the same as for the parent product. Currently, we can solve this on SQL level with doing something like ``COALESCE(product.attendee_name, parent.attendee_name)``. The same goes for inheriting names from invoice addresses, which we use in some places. This will get a lot more complicated, since we can't use COALESCE any more, but instead need to use more complicated ``IF()`` clauses to check if all name fields are ``NULL`` on the current row.

## Rejected designs
* Store names as a JSON field. This would solve most of our problems, but only on PostgreSQL, and we're not planning to go PostgreSQL only.

**Comment by stapelberg** [comment]
I think the “calling name” suggestion is the right track, and can be extended to other purposes: http://codel10n.com/first-middle-last-why-not-full-name/ outlines how e.g. Google Contacts allows you to specify a “name for sorting”. Maybe the “name for sorting” could be auto-filled depending on the shop’s locale.

In summary, using one field per purpose (sorting, badge, salutation, …) seems cleanest to me.

**Comment by mdelete** [comment]
I like the idea with the "name for sorting", because sorting is the only thing, I think, where this is crucial. Currently, in our typical scenario, having one name field will effectively lead to sorting for first-names, which is not optimal with the last-name being more distinct.

As Germany is moderately title-crazy many people would but their "Dr." in front in the single name field, which leads to many people in the "D" section with not much more to go on. So the "name for sorting" would also eliminate the need for a title field.

A "name for sorting" **must** be auto-filled, having this in a user-filled form would be terrible UX, because no one would understand it. **But** I have no idea how a form should look like to make this happen.

**Comment by stapelberg** [comment]
> A "name for sorting" must be auto-filled, having this in a user-filled form would be terrible UX, because no one would understand it. But I have no idea how a form should look like to make this happen.

Maybe the field should be entirely invisible in locales where the rules are clear-cut enough that one can do that.

**Comment by raphaelm** [comment]
> > A "name for sorting" must be auto-filled, having this in a user-filled form would be terrible UX, because no one would understand it. But I have no idea how a form should look like to make this happen.
> Maybe the field should be entirely invisible in locales where the rules are clear-cut enough that one can do that.

I'm not sure there are any locales where this is possible. Trying to build a name-cutting algorithm (even just for autofilling a field) seems like the best and the worst idea at the same time. Locales don't help you that much, since even in Germany, you have people with e.g. Brasilian names living here. This will likely never work well enough and will be lots and lots of work.

That's also why I think a field "name for sorting" is not a great idea. :( It works for a "Contacts" app, since the person entering the data is the same person using the sorted list later. Here, it's different: Attendees input the data, and organizers later need to find it. Attendees won't understand it and shouldn't be bothered to learn about it or even to enter parts of their name twice. Also, organizers might want to decide later by which name part they want to sort by.

**Comment by raphaelm** [comment]
> I think splitting names in the German way, even if the name is Brasilian, is working as intended: the event organizers are telling the software how they want names to be split, and doing anything else (e.g. detecting Brasilian names and splitting them as per the Brasilian name-splitting rules) is unexpected behavior.

Agreed.
However, how would splitting German names even work? Splitting with the last space? First space? Even among the German contributors of this project, there are both people who totally use their middle name everywhere ("Jan Felix Wiebe", @jfwiebe) and people with a last name consisting of two words ("Jonas Große Sundrup", @cherti). I don't see how this can be detected automatically, without asking the user to split – and asking the user for "given name" and "family name" will confuse them much less than "full name" and "full name for sorting", or even "full name" and "family name (again)".

**Comment by stapelberg** [comment]
That’s fair.

I think if you want to fully support this feature across locales, you’ll need separate database tables and form fields for each logical part of a name.

If an approximation is good enough, sorting by first word or last word should cover most use-cases. People who have multiple name parts are probably used to looking in both places…

**Comment by mdelete** [comment]
Splitting names is really complicated and never works at a level that is usable in my opinion, one wrongly split name can really get you a headache at the check-in. To mention my earlier example with the title in front of the name, there are occasions where the check-in just has a sorted list: Going by first-name is inconvenient (for Germany at least) but works, going by last-name works well. If the list contains titles with high precedence in the sorting-key, or if you split a name wrongly, it leads to names that are almost impossible to find even on short lists. And at the worst possible time, at the checkin.

So for a quick win, it would be nice to have a name-for-sorting field that has the common german/austrian titles stripped. I think in other countries they consider a title not to be part of a name. All other auto-fill algorithms can lead to serious errors.

Maybe preview functions when the attendee fills out the name field would be nice: "This is what you badge would look like", and: "This will be your name on the attendee-list" ...

**Comment by cherti** [comment]
> Maybe preview functions when the attendee fills out the name field would be nice: "This is what you badge would look like", and: "This will be your name on the attendee-list" ...

And then a "Customize"-Button or so if the attendee wants the badge to look differently? With only after pressing the "Customize"-button showing additional customizable (prefilled) fields?


I don't think that splitting automatically based on one Text-Input-Field is possible at all. Taking the example of "Jan Felix Wiebe" and "Jonas Große Sundrup", both are three word expressions, but would have to be split differently. This couldn't at most be split heuristically (by forename-dictionary or something maybe), but then again, without prior knowledge, no one guarantees that "Felix" is actually a second forename as I know people whose last name is the same as other people's forename, making it impossible to automatically deal with all cases.

What could be done, though, would be a one-text-field that splits, but offers the user the possibility to correct. Then again simply offering multiple fields to enter in the first place is probably much more simple.

**Comment by johan12345** [comment]
> What could be done, though, would be a one-text-field that splits, but offers the user the possibility to correct. Then again simply offering multiple fields to enter in the first place is probably much more simple.

There seem to be at least 2 python libraries that do this splitting for western names, [one](https://github.com/derek73/python-nameparser) using hard-coded information about prefixes, titles, etc., and [one](https://github.com/datamade/probablepeople) using machine learning (natural language processing) trained on common first and last names. As said above, both will never be perfect (none of them parses "Jonas Große Sundrup" correctly, but at least things like "Mr John Doe" or "Dr. John Doe" work), so this:

> Maybe preview functions when the attendee fills out the name field would be nice: "This is what you badge would look like", and: "This will be your name on the attendee-list" ...

sounds like a good way to allow the user to correct the result.

**Comment by Problemkerze** [comment]
From the point of view of an international event, we have experienced that the participants are able to split their name into first and last name, but that everyone has a different way of doing it. (e.g. [call name] - [second first name, last name] vs. [call name, second first name] - [last name])
As long as the respective participant is satisfied, so are we.
Basically, we are happy with recording the complete name in one field, especially when using the check-in feature. With two shortcomings:

We want to print badges and it would be nice to be able to print the last name in a separate line. (Besides, there doesn't seem to be a line break in the badges, but that's another topic.)

Also, some participants insist on having their name part "Dr.", which should not be part of any sort. I don't think anybody wants "Prof. Dr. Ralf Müller" to be sorted under P. (At our current event we already have two Dr.'s among the twelve registered participants.)

A rather minor matter: In my private environment I have events in which most of the participants use nicknames and some of them are only known by them (Bündische Jugend in Deutschland - Pfadfinder, Waldjugend, etc.). Maybe this phenomenon also exists in other groups.
